### DISCUSSION TITLE
Install-code support for ZStack 1.2 question

### DISCUSSION THREAD

**Discussion by Switch123456789** [Koenkk/zigbee2mqtt#19176] [Devices]
Hi everyone,
today I bought a `Bosch BTH-RA` because according to [zigbee2mqtt.io](https://www.zigbee2mqtt.io/devices/BTH-RA.html) it is compatible and according to [Texas Instruments](https://www.ti.com/product/CC2531) my CC2531 based USB-Dongle with zstack 1.2 supports Zigbee 3. However, if I try to add the install-code I get an error:

`Request 'zigbee2mqtt/bridge/request/install_code/add' failed with error: 'Install code is not supported for ZStack 1.2 adapter'`

Now I am wondering what exactly is the reason for that? Is it a limitation of z2m or a limitation of the Zstack-firmware? Or a hardware-limitation? Or is it because install-codes are not part of the zigbee 3 specification?

I feel a bit lost in the zigbee device-jungle now. I double & tripple check everything and still end up buying incompatible devices. I spent an hour googling and couldn't find information about which coordinator-devices/chips/firmwares support install-codes. If anyone could shed some light on this for me I would rather pay him/her instead of wasting more money on incompatible devices.

**Comment by Switch123456789** [comment]
I am considering to upgrade my Coordinator/Bridge, any hints to wich one supports install-codes would be appreciated.

**Reply by lux73** [reply]
AFAIK CC2531 is zigbee 1.2 only

and any CC2562x based Coordinator should support Install Codes

https://github.com/Koenkk/zigbee2mqtt/issues/14510

**Reply by Switch123456789** [reply]
On TI's website it says `Protocols: RF4CE, Zigbee 3.0` (see OP).
However, I saw on the ZStack-Git that there is a newer firmware with Zigbee 3 support but it must be pretty unreliable and I would have to buy 2 adapters to flash it, which cost about the same as another Zigbee 3 dongle :laughing:
But the weirdest thing is, it is paired to a zigbee 3 device right now! This is even more confusing.

Thank you @lux73 for clarifying! I guess I just have to look at the firmware capabilities, not just the chip-specs before I buy.
